### PR TITLE
EX-1390 & EX-1393 Write ActBlue class for Parsons.

### PR DESCRIPTION
**Pull request by torimarbois**
This pull request includes the initialization of the ActBlue connector class, which currently includes a post_request method, a get_download_url method, and a poll_for_download_url method. You'll notice the DocStrings at the beginning of each method (and the parent class) aren't filled in - we didn't get to that as part of this ticket. 
We also wrote a __main__ function for testing purposes, which displays the function calls in the order they should be made. 

We worked closely off of a Ruby example script that shows how to interact with the CSV API, which can be found in the 'CSV Download API: Internal Documentation' in Confluence.

### CODE REVIEW

**Comment by torimarbois** [comment]
Note: we also had to make changes to the accepted status_codes within the post_request method of the Parsons api_connector class (we added 202 as a successful code, since that is what the AB CSV API returns after successful POST requests). Those changes are NOT reflected in this PR and therefore, if you try to run this code locally as-is, it will not work.

File: parsons/__init__.py [hunk 1]
```diff
@@ -98,7 +99,8 @@
         'Sisense',
         'SurveyGizmo',
         'Alchemer',
-        'Quickbase'
+        'Quickbase',
+        'ActBlue'
```
Comment (eudaimonious) [inline parsons/__init__.py:103]: On line 10 above, you'll notice that users can set an environment variable to PARSONS_SKIP_IMPORT_ALL and that will skip this block of code, including the definition of __all__ on line 57. To get our ActBlue connector working if the user has opted not to install all the other packages, we need to do our import and define __all__ in the actblue-specific init file.

File: parsons/actblue/actblue.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+class ActBlue(object):
+    """
+    Instantiate class.
+
+       `Args:`
+    """
+
+    def __init__(self, actblue_client_uuid=None, actblue_client_secret=None):
+        self.actblue_client_uuid = check_env.check('ACTBLUE_CLIENT_UUID', actblue_client_uuid)
+        self.actblue_client_secret = check_env.check('ACTBLUE_CLIENT_SECRET', actblue_client_secret)
```
Comment (eudaimonious) [inline parsons/actblue/actblue.py:21]: I noticed that this got checked in to github on an earlier commit (and so is still discoverable) please delete that credential in our app. In the web app, please navigate to the page where you created this credential, remove it, and create a new one to use instead, if you haven't already.

File: parsons/actblue/actblue.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        `Args:`
+        """
+
+        body = {
+            "csv_type": csv_type,
+            "date_range_start": date_range_start,
+            "date_range_end": date_range_end
+        }
+
+        response = self.client.post_request(url="csvs", json=body)
```
Comment (eudaimonious) [inline parsons/actblue/actblue.py:42]: I think we need to add in some error handling here. What should happen if, for example, someone misspells their csv type as "payd_contributionz." Should we abort the rest of the script and print out what the server's error response was?

File: parsons/actblue/actblue.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        """
+        Instantiate class.
+
+        `Args:`
+        """
+
+        try: 
+            response = self.client.get_request(url=f"csvs/{csv_id}")
+            download_url = response["download_url"]
+        except requests.exceptions.HTTPError:
```
Comment (eudaimonious) [inline parsons/actblue/actblue.py:55]: So if on polling we receive an error we'll try again a second later. That could be good for a transient error, but I'm not sure what would be going on to cause that scenario... I feel like we might want to come back to this.

File: parsons/actblue/actblue.py (hunk 2 above)
Comment (crystaljg) [reply parsons/actblue/actblue.py:21]: I removed the credentials for the account and will work on scrubbing the git history, apologies for not doing this sooner!

Comment (crystaljg) [reply parsons/actblue/actblue.py:21]: I rebased the commit history to remove these, let me know if there's anything that needs to happen on this! @torimarbois I can show you what I did to edit the commit history tomorrow if you haven't done it before

File: parsons/actblue/actblue.py (hunk 3 above)
Comment (crystaljg) [reply parsons/actblue/actblue.py:42]: Turns out parsons api_connector does this for us!

File: parsons/__init__.py (hunk 1 above)
Comment (eudaimonious) [reply parsons/__init__.py:103]: @torimarbois I think we forgot to get to this bit when pairing today

**Comment by torimarbois** [comment]
The past 3 commits address EX-1393, which covered finishing up the ActBlue class and filling out the Doc strings. I accidentally pushed the final commit to this branch instead of a new one, hence why this is still linked to the first ticket/MR. Please review!

File: parsons/actblue/actblue.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        post_request_response = self.post_request(csv_type, date_range_start, date_range_end)
+        csv_id = post_request_response["id"]
+        download_url = self.poll_for_download_url(csv_id)
+        table = Table.from_csv(download_url)
+        return table
+
+def main():
+    actblue = ActBlue()
+    table = actblue.get_contributions("paid_contributions", "2020-01-01", "2020-02-01")
```
Comment (sonjapeterson) [inline parsons/actblue/actblue.py:129]: I see the dates and types are hard coded here - is that just temporary?

**Review by sonjapeterson — APPROVED** [review]
This looks great! one minor q

File: parsons/actblue/actblue.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                The ActBlue provided Client UUID. Not required if ``ACTBLUE_CLIENT_UUID`` env variable set.
+            actblue_client_secret: str
+                The ActBlue provided Client Secret. Not required if ``ACTBLUE_CLIENT_SECRET`` env variable set.
+
+        For instructions on how to generate a Client UUID and Client Secret set, visit https://secure.actblue.com/docs/csv_api#authentication.
+    """
+
+    def __init__(self, actblue_client_uuid=None, actblue_client_secret=None):
+        self.actblue_client_uuid = check_env.check('ACTBLUE_CLIENT_UUID', actblue_client_uuid)
+        self.actblue_client_secret = check_env.check('ACTBLUE_CLIENT_SECRET', actblue_client_secret)
```
Comment (sonjapeterson) [inline parsons/actblue/actblue.py:26]: could we / should we raise an error if these are missing?

File: parsons/actblue/actblue.py (hunk 5 above)
Comment (torimarbois) [reply parsons/actblue/actblue.py:129]: Yes! The hard coded parameters were just used for testing.

**Comment by eudaimonious** [comment]
A recent example of some test data in the repo, might be something we want to emulate... https://github.com/move-coop/parsons/pull/598/files#diff-3e982cdc84d866feb15fc7af6c57dca9a7fb0ef06de4d45f70b56a31c4419368

**Comment by polirex** [comment]
This PR looks pretty straightforward to me, however I wanted to run some tests before I approved it and while running the script I ran into a bug in the `get_contributions` method for the `csv_id` variable. The error returns a `'NoneType' object is not subscriptable`

**Comment by crystaljg** [comment]
> This PR looks pretty straightforward to me, however I wanted to run some tests before I approved it and while running the script I ran into a bug in the `get_contributions` method for the `csv_id` variable. The error returns a `'NoneType' object is not subscriptable`

This was resolved by doing `pip3 install -e .` to install the version of parsons where we added `204` as an accepted response in parson's `api_connector`. We're going to add a more useful error around this and see if we have to do anything special to make sure this isn't a problem when we submit a PR to the repo.

File: parsons/actblue/actblue.py (hunk 6 above)
Comment (torimarbois) [reply parsons/actblue/actblue.py:26]: yes! Just added a TODO to the code; we will add a 'raise error' line before the final MR.

File: parsons/actblue/actblue.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        self.uri = 'https://secure.abstaging.com/api/v1/' #TODO: change to prod link
```
Comment (sonjapeterson) [inline parsons/actblue/actblue.py:27]: let's make sure to change this to prod before merging

File: parsons/__init__.py (hunk 1 above)
Comment (torimarbois) [reply parsons/__init__.py:103]: This was added in a commit to the actblue_class_testing branch.

File: parsons/actblue/actblue.py (hunk 4 above)
Comment (torimarbois) [reply parsons/actblue/actblue.py]: This was changed!

File: parsons/actblue/actblue.py (hunk 7 above)
Comment (torimarbois) [reply parsons/actblue/actblue.py:27]: We'll make sure to make this change to include with the final MR to the parent parsons repo.

**Pull request closed without merging**